botbot_plugins/plugins/vote.py

```diff
@@ -0,0 +1,394 @@
+    use the “countvotes”.
+
+    Finally, to end a vote, use the “endvote” command.
+    """
+
+    config_class = Config
+
+    @listens_to_regex_command("startvote",
+                              ur"(?P<name>[^\[\]]+?)?\s*(\[(?P<options>.*)\])?\s*$")
+    def startvote(self, line, name, options):
+        options_sep = self.config["options_separator"]
+        if self.retrieve("votes"):
+            return (ERROR_MESSAGES["voting_already_running"].
+                    format(author=line.user))
+
+        default_options = [DEFAULT_VOTE_OPTION]
+        options =\
+            filter(bool, [option_name.strip() for option_name
+                          in (options.split(options_sep) if options else [])])
+        options = options or default_options
+        # Abstaining should always be an option
+        options.append(ABSTAIN_OPTION)
+        options = list(set(options))
+
+        # NOTE: this will create a for and against in __abstain__,
+        # but for simplicity's sake we won't try and prevent this.
+        # All abstainers will be added to __abstain__.for.
+        initial_votes = {
+            option_name: {FOR_NAME: [], AGAINST_NAME: []}
+            for option_name in options
+        }
+
+        self.store("name", json.dumps(name))
+        self.store("options", json.dumps(options))
+        self.store("votes", json.dumps(initial_votes))
+        self.store("changes_since_countvote", json.dumps(False))
+
+        return (INFO_MESSAGES["voting_started_name"].format(name=name)
+                if name else INFO_MESSAGES["voting_started"])
+
+    def _depends_on_votestarted(throw_error=True):
+        def decorator(func):
+            @wraps(func)
+            def wrap(*args, **kwargs):
+                slf = args[0]
+                if slf.retrieve("votes"):
+                    return func(*args, **kwargs)
+                elif throw_error:
+                    return ERROR_MESSAGES["no_vote_started"]
+                return ""
+            return wrap
+        return decorator
+
+    @listens_to_command("endvote")
+    @_depends_on_votestarted()
+    def endvote(self, line, args):
+        name = json.loads(self.retrieve("name"))
+        changes_since_countvote = json.loads(
+            self.retrieve("changes_since_countvote"))
+        reply = (INFO_MESSAGES["voting_ended_name"].format(name=name)
+                 if name else INFO_MESSAGES["voting_ended"])
+        if changes_since_countvote:
+            reply += "\n" + self._print_votes()
+
+        self.delete("name")
+        self.delete("options")
+        self.delete("votes")
+        self.delete("changes_since_countvote")
+
+        return reply
+
+    @listens_to_command("countvotes")
+    @_depends_on_votestarted()
+    def countvotes(self, line, args):
+        self.store("changes_since_countvote", json.dumps(False))
+        return self._print_votes()
+
+    vote_regex_template = (
```

Comment:
Also, very anti-python to use such a non verbose, almost functional way of writing code. I get the allure of shorter code but it should be readable. 

See BDFL's view on functional programming in python - https://developers.slashdot.org/story/13/08/25/2115204/interviews-guido-van-rossum-answers-your-questions

Reply:
What I meant here is the fact that there is so much going on in one line. You could instead flesh it out to make the regex more readable.
You can use format strings for the intial and last part of the regex and then substitute the middle dynamically built string with a variable name like `escaped_voting_shortcut_pattern` or something along that line.

It might look like

```
u"^((?P<vote>({escaped_voting_shortcut_pattern}\\\\){vote_modifier})(?P<option>.*))$")
```

Reply:
```python
x = "{name} test {{later}}".format(name="name")
x = x.format(later="late")
>>> print(x)
>>> 'name test late'
```